### PR TITLE
GAWB-320 Update Clone Workspace to handle Realms

### PR DESCRIPTION
**Pull request by jmthibault79**
Also add Create and Clone Workspace tests with Realms
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] Tell  that the PR exists if he wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [x] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x]  sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Ask QA if it's a convenient time to merge this PR**; hold off if they're in the middle of testing something
- [x] **Submitter**: Merge to develop 
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: Check configuration files in Jenkins in case they need changes
- [x] **Submitter**: Verify swagger UI on dev server still works after deployment
- [x] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [x] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 1]
```diff
@@ -265,22 +265,27 @@ class WorkspaceService(protected val userInfo: UserInfo, dataSource: DataSource,
...
       withWorkspaceContextAndPermissions(sourceWorkspaceName,WorkspaceAccessLevels.Read,txn) { sourceWorkspaceContext =>
-        withNewWorkspaceContext(
-          WorkspaceRequest(
-            namespace = destWorkspaceName.namespace,
-            name = destWorkspaceName.name,
-            realm = sourceWorkspaceContext.workspace.realm,
-            attributes = sourceWorkspaceContext.workspace.attributes),txn) { destWorkspaceContext =>
+
+        // add to or replace current attributes, on an individual basis
+        val newAttrs = sourceWorkspaceContext.workspace.attributes ++ destWorkspaceRequest.attributes
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: This was not explicitly asked for, but a WorkspaceRequest comes with attributes, and this seemed like the most reasonable way to handle them.  I added tests for this behavior as well.

An alternative would be adding a new WorkspaceCloneRequest case class without attributes.

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala [hunk 2]
```diff
@@ -199,6 +199,41 @@ class WorkspaceApiServiceSpec extends ApiServiceSpec {
       }
   }
 
+  it should "create a workspace with a Realm" in withTestDataApiServices { services =>
```
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala:202]: Also not part of GAWB-320, but I think it's a necessary addition to GAWB-285.

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala [hunk 3]
```diff
@@ -354,6 +389,105 @@ class WorkspaceApiServiceSpec extends ApiServiceSpec {
...
+        assertResult(StatusCodes.Created) {
+          status
+        }
+        assertResult(workspaceWithRealm.realm) {
+          responseAs[Workspace].realm
+        }
+      }
+  }
+
+  it should "set the Realm when cloning a workspace with no Realm" in withTestDataApiServices { services =>
```
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/WorkspaceApiServiceSpec.scala]: happy to split this in two if you prefer

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 4]
```diff
@@ -265,22 +265,27 @@ class WorkspaceService(protected val userInfo: UserInfo, dataSource: DataSource,
...
+
+        // if the source has a realm, the dest must also have that realm.  Otherwise, the caller chooses
+        val newRealm = sourceWorkspaceContext.workspace.realm match {
+          case Some(realm) => Some(realm)
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: should we throw an error if both the source and destination define a realm that are different?

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 1 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: this is probably fine, there are certainly use cases out there where we might want to overwrite a workspace attribute on the clone, e.g. description, maybe tell @MattPutnam about this.

**Comment by dvoet** [comment]
one thing on throwing an exception then :+1:

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 4 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: That would mean the UI/Orch messed up.  So yes.

**Pull request merged**